File: hdk/node/src/plugin/context.ts
import { HookContext, TenantContext } from "../protocol/messages";
import { HookName } from "../protocol/constants";

export class Context {
  readonly hook: HookName;
  readonly requestId: string;
  readonly orgId?: string;
  readonly projectId?: string;
  readonly environmentId?: string;
  readonly metadata: Readonly<Record<string, string>>;

  constructor(protoCtx: HookContext) {
    this.hook = protoCtx.hook;
    this.requestId = protoCtx.request_id;
    this.orgId = protoCtx.tenant?.organization_id;
    this.projectId = protoCtx.tenant?.project_id;
    this.environmentId = protoCtx.tenant?.environment_id;
    this.metadata = Object.freeze({ ...(protoCtx.metadata || {}) });
    Object.freeze(this);
  }

  // Backward compatibility getters
  get organization_id(): string | undefined {
    return this.orgId;
  }
  get project_id(): string | undefined {
    return this.projectId;
  }
  get environment_id(): string | undefined {
    return this.environmentId;
  }
  get request_id(): string {
    return this.requestId;
  }
}

File: hdk/go/plugin/context.go
package plugin

import "github.com/bhakha-services/naagmani-plugins/hdk/go/protocol"

// TenantContext carries multi-tenant scoping information for the request.
type TenantContext struct {
	OrganizationID string
	ProjectID      string
	EnvironmentID  string
}

// Context carries metadata and tenant information passed to plugin hooks.
type Context struct {
	Hook      string
	RequestID string
	Tenant    TenantContext
	Metadata  map[string]string
}

func contextFromProtocol(ctx *protocol.HookContext) *Context {
	c := &Context{
		Hook:      ctx.Hook,
		RequestID: ctx.RequestID,
		Tenant: TenantContext{
			OrganizationID: ctx.Tenant.OrganizationID,
			ProjectID:      ctx.Tenant.ProjectID,
			EnvironmentID:  ctx.Tenant.EnvironmentID,
		},
		Metadata: make(map[string]string),
	}
	for k, v := range ctx.Metadata {
		c.Metadata[k] = v
	}
	return c
}
